doc_gen.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+declare(strict_types = 1);
+
+function generateNamespaceLine($namespace)
+{
+    static $last_namespace = null;
+
+    if ($namespace !== $last_namespace) {
+        $last_namespace = $namespace;
+        return "\n" . $namespace . "  \n";//double space for new line
+    }
+
+    return '';
+}
+
+function generateLink($version, $filename, $startLine): string
+{
+    $string = "https://github.com/azjezz/psl/blob/%s/src/%s#L%d";
+
+    return  sprintf(
+        $string,
+        $version,
+        $filename,
+        $startLine
+    );
+}
+
+function normaliseFilename($filename): string
+{
+    $lastPos = strrpos($filename, 'psl/src');
+    if ($lastPos !== false) {
+        $filename = substr($filename, $lastPos + strlen('psl/src/'));
+    }
+    return $filename;
+}
+
+function generateFunctionLine($sha, $function): string
+{
+    $rf = new ReflectionFunction($function);
+
+    $filename = $rf->getFileName();
+    $filename = normaliseFilename($filename);
+
+    $link = generateLink($sha, $filename, $rf->getStartLine());
+    $rf = new ReflectionFunction($function);
+
+    $line = generateNamespaceLine($rf->getNamespaceName());
+
+    $functionWithoutNamespace = str_replace($rf->getNamespaceName(), '', $function);
+    $functionWithoutNamespace = ltrim($functionWithoutNamespace, '\\');
+
+    $line .= sprintf(
+        "* <a href='%s'>%s</a>  ", //double space for new line
+        $link,
+        $functionWithoutNamespace
+    );
+
+    return $line;
+}
+
+function generateInterfaceLine($sha, $interface): string
+{
+    $rc = new ReflectionClass($interface);
+
+    $filename = $rc->getFileName();
+    $filename = normaliseFilename($filename);
+
+    $link = generateLink($sha, $filename, $rc->getStartLine());
+
+    $line = generateNamespaceLine($rc->getNamespaceName());
+
+    $functionWithoutNamespace = str_replace($rc->getNamespaceName(), '', $interface);
+    $functionWithoutNamespace = ltrim($functionWithoutNamespace, '\\');
+
+    $line .= sprintf(
+        "* <a href='%s'>%s</a>  ", //double space for new line
+        $link,
+        $functionWithoutNamespace
+    );
+
+    return $line;
+}
+
+function generateDocList(): string
+{
+    $version = '1.5.x';
```

Comment:
since this will always be ran against the default branch ( upcoming release ), i think we can use the API to retrieve the default branch.

```suggestion
    $repository = Json\typed(file_get_contents('https://api.github.com/repos/azjezz/psl'), Type\dict(
      Type\string(),
      Type\mixed(),
    ));

    $version = Json\string()->assert($repository['default_branch']);
```

also `use Psl\Json` and `use Psl\Type` :)

however, we might wanna run this against other branches, so we can probably add `--sha` argument for it in the future.